third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp

```diff
@@ -59,82 +60,172 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
 
+  // Per-dimension shape/offset checks
+  auto srcShape = srcTy.getShape();
+  auto dstShape = dstTy.getShape();
+  auto offsets = getStaticOffsets();
   auto shapePerCTATile = mlir::triton::gpu::getShapePerCTATile(srcTy);
-  shapePerCTATile[0] =
-      std::min(static_cast<unsigned>(srcShape[0]), shapePerCTATile[0]);
-  shapePerCTATile[1] =
-      std::min(static_cast<unsigned>(srcShape[1]), shapePerCTATile[1]);
-  if (sizes[0] % shapePerCTATile[0] != 0 ||
-      sizes[1] % shapePerCTATile[1] != 0) {
-    return emitError() << "sizes [" << sizes
-                       << "] must be a multiple of shapePerCTATile ["
-                       << shapePerCTATile << "]";
-  }
+  size_t rank = srcShape.size();
 
-  if (offsets[0] % shapePerCTATile[0] != 0 ||
-      offsets[1] % shapePerCTATile[1] != 0) {
-    return emitError() << "offset [" << offsets
-                       << "] must be a multiple of shapePerCTATile ["
-                       << shapePerCTATile << "]";
+  auto failDim = [&](StringRef msg, int i) -> LogicalResult {
+    return emitError(msg) << " at dimension " << i;
+  };
+
+  for (size_t i = 0; i < rank; ++i) {
+    if (dstShape[i] > srcShape[i])
+      return failDim("result shape cannot exceed source shape", i);
+    if (offsets[i] + dstShape[i] > srcShape[i])
+      return failDim("invalid offset", i);
+    if (dstShape[i] % shapePerCTATile[i] != 0)
+      return emitError("result shape must be multiple of shapePerCTATile");
+    if (offsets[i] % shapePerCTATile[i] != 0)
+      return emitError("offset must be multiple of shapePerCTATile");
   }
 
+  // Verify that source and destination layout match on a CTA tile.
+  if (!matchingLayoutsOnCTATile(getContext(), srcTy, dstTy,
+                                [&](const Twine &msg) { emitError() << msg; }))
+    return failure();
+
   return success();
 }
 
+struct CononicalizeExtractSliceAndConcat
+    : public mlir::OpRewritePattern<amdgpu::ExtractSliceOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(amdgpu::ExtractSliceOp op,
+                  PatternRewriter &rewriter) const override {
+    auto concatOp = op.getSource().getDefiningOp<amdgpu::ConcatOp>();
+    if (!concatOp)
+      return failure();
+
+    auto offset = op.getStaticOffsets();
+
+    auto sliceResult = op.getResult();
+    auto sliceResultType = sliceResult.getType();
+    RankedTensorType dstType =
+        cast<RankedTensorType>(concatOp.getResult().getType());
+    auto dstShape = dstType.getShape();
+
+    auto concatItem = concatOp.getSources().front();
+    auto concatItemType = dyn_cast<RankedTensorType>(concatItem.getType());
+    if (!concatItemType)
+      return failure();
+
+    if (sliceResultType != concatItemType)
+      return failure();
+
+    auto srcShape = concatItemType.getShape();
+    auto rank = srcShape.size();
+    std::vector<unsigned> defaultOrder(rank);
+    std::iota(defaultOrder.rbegin(), defaultOrder.rend(), 0);
+
```

Comment:
Add some comments to explain the following logic given it's not very straightforward?